File: solution/java/day21/src/main/java/audit/AddRecordUseCase.java
package audit;

import java.io.IOException;

public class AddRecordUseCase {
    private final String directory;
    private final AuditManager auditManager;
    private final Persister persister;

    public AddRecordUseCase(String directory, AuditManager auditManager, Persister persister) {
        this.directory = directory;
        this.auditManager = auditManager;
        this.persister = persister;
    }

    public void handle(AddNewVisitor addNewVisitor) throws IOException {
        var files = persister.readDirectory(directory);
        var update = auditManager.addRecord(files, addNewVisitor.visitor(), addNewVisitor.time());

        persister.applyUpdate(directory, update);
    }
}

File: solution/kotlin/day21/src/main/kotlin/audit/AddRecordUseCase.kt
package audit

class AddRecordUseCase(
    private val directory: String,
    private val auditManager: AuditManager,
    private val persister: Persister
) {
    fun handle(addNewVisitor: AddNewVisitor) {
        val files = persister.readDirectory(directory)
        val update = auditManager.addRecord(files, addNewVisitor.visitor, addNewVisitor.time)

        persister.applyUpdate(directory, update)
    }
}